File: LeetCode/LeetCode 77. 组合 (dfs).cpp
class Solution {
public:
    vector <int> ss;
    vector <vector<int>> ans;

    void dfs(int x, int n, int k) {
        if(ss.size() == k) {
            ans.push_back(ss);
            return ;
        }
        for(int i = x; i <= n; i++) {
            ss.push_back(i);
            dfs(i + 1, n, k);
            ss.pop_back();
        }
        return ;
    }

    vector<vector<int>> combine(int n, int k) {
        dfs(1, n, k);
        return ans;
    }
};

File: LeetCode/LeetCode 77. 组合 (dfs).py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans = list()
        ss = list()

        def dfs(x: int, n: int, k: int):
            if len(ss) == k:
                ans.append(ss[:])   #浅拷贝
                return 
            
            for i in range(x, n + 1):
                ss.append(i)
                dfs(i + 1, n, k)
                ss.pop()
        
        dfs(1, n, k)
        return ans